reduce the number of allocations in the WatchServer during objects serialisation

p0lyn0mial opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

The WatchServer is largely responsible for streaming data received from the storage layer. It turns out that sending a single event per consumer requires **4** memory allocations, visualized in the following image. Two of which deserve special attention, namely allocations 1 and 3 because they won't reuse memory and rely on the GC for cleanup. In other words, the more events we need to send, the more (temporary) memory will be used. In contrast, the other two allocations are already optimized they reuse memory instead of creating new buffers for every single event.

For better memory utilization, this PR changes the protobuf encoders to accept a memory allocator and changes the WatchServer to allocate **a single buffer** (combines 1, 3 and 4) for the entire watch session and pass it to encoders during objects serialization.

![watch_server_allocs](https://user-images.githubusercontent.com/21304151/155290189-8a076702-9b9f-4f26-8bf9-1720f5f04db4.png)

I am attaching the results from the benchmarks included in this PR that show an improvement.

The results for protobuf.Serializer
```
BenchmarkProtobufEncoder/an_obj_with_1kB_payload-12         	  499345	      2270 ns/op	    1192 B/op	       3 allocs/op
BenchmarkProtobufEncoder/an_obj_with_10kB_payload-12        	  210517	      6327 ns/op	   10280 B/op	       3 allocs/op
BenchmarkProtobufEncoder/an_obj_with_100kB_payload-12       	   27073	     41799 ns/op	  106536 B/op	       3 allocs/op
BenchmarkProtobufEncoder/an_obj_with_1MB_payload-12         	    2787	    372108 ns/op	 1007658 B/op	       3 allocs/op

BenchmarkProtobufEncodeWithAllocator/an_obj_with_1kB_payload-12         	  800178	      1512 ns/op	      40 B/op	       2 allocs/op
BenchmarkProtobufEncodeWithAllocator/an_obj_with_10kB_payload-12        	  719036	      1740 ns/op	      40 B/op	       2 allocs/op
BenchmarkProtobufEncodeWithAllocator/an_obj_with_100kB_payload-12       	  201190	      5908 ns/op	      40 B/op	       2 allocs/op
BenchmarkProtobufEncodeWithAllocator/an_obj_with_1MB_payload-12         	   16674	     73044 ns/op	     100 B/op	       2 allocs/op
```

The results for protobuf.RawSerializer:
```
BenchmarkRawProtobufEncoder/an_obj_with_1kB_payload-12                  	  669680	      1978 ns/op	    1192 B/op	       3 allocs/op
BenchmarkRawProtobufEncoder/an_obj_with_10kB_payload-12                 	  188064	      6155 ns/op	   10280 B/op	       3 allocs/op
BenchmarkRawProtobufEncoder/an_obj_with_100kB_payload-12                	   29367	     41180 ns/op	  106536 B/op	       3 allocs/op
BenchmarkRawProtobufEncoder/an_obj_with_1MB_payload-12                  	    3354	    370200 ns/op	 1007656 B/op	       3 allocs/op

BenchmarkRawProtobufEncodeWithAllocator/an_obj_with_1kB_payload-12      	 1000000	      1128 ns/op	      40 B/op	       2 allocs/op
BenchmarkRawProtobufEncodeWithAllocator/an_obj_with_10kB_payload-12     	  785414	      1377 ns/op	      40 B/op	       2 allocs/op
BenchmarkRawProtobufEncodeWithAllocator/an_obj_with_100kB_payload-12    	  209011	      5314 ns/op	      40 B/op	       2 allocs/op
BenchmarkRawProtobufEncodeWithAllocator/an_obj_with_1MB_payload-12      	   22004	     46844 ns/op	      85 B/op	       2 allocs/op
```

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

https://github.com/kubernetes/enhancements/issues/3157

#### Special notes for your reviewer: 

you will find more info in https://github.com/kubernetes/enhancements/pull/3142

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/3142
```

p0lyn0mial committed:
update the watch server to use EncoderWithAllocator during object serialization.

It allows us to allocate a single buffer for the entire watch session and release it when a watch connection is closed.
Previously memory was allocated for every object serialization putting a lot of pressure on GC and consuming more memory than needed.

p0lyn0mial committed:
codec interfaces

p0lyn0mial committed:
codec: exposes EncodeWithAllocator method

The new method is implemented by the protobuf serializer and helps to reduce memory footprint during object serialization.

p0lyn0mial committed:
provides EncodeWithAllocator method for the protobuf encoder

The new method allows for providing a memory allocator for efficient memory usage during object serialization.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/allocator.go:
```diff
@@ -0,0 +1,21 @@
+	buf []byte
+}
+
+var _ runtime.MemoryAllocator = &Allocator{}
```
+1

1 buffer will not be enough anyway, because we may be serializing many different objects at once.
Using sync.Pool will allow us to utilize that nicely.

p0lyn0mial replied:
good idea, I like it, thanks for pointing it out.

p0lyn0mial replied:
I think this should do the trick - https://github.com/kubernetes/kubernetes/pull/108186/commits/6e3c32adf7fe8ab7910f30f7e301255ba57b62e2

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/testing/types.go:
```diff
@@ -0,0 +1,17 @@
+
+// TypeA
+// +protobuf=true
+type TypeA struct {
```
Do we need to introduce new types really?

Can't we reuse existing ones, e.g.:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/testapigroup/v1/types.go#L33

p0lyn0mial replied:
oh, I missed `testapigroup` package, no need to intro a new type then.

wojtek-t replied:
Please remove then - it will make this PR much smaller (and thus easier to review).

p0lyn0mial replied:
done.

wojtek-t reviewed (commented):
Just two quick comments - I will take a deeper look in the upcoming days, but I think we should address them before that.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/allocator.go:
```diff
@@ -0,0 +1,42 @@
+var AllocatorPool = sync.Pool{
+	New: func() interface{} {
+		return &Allocator{}
+	},
```
One more thing I was thinking about is whether we want a single pool, or maybe we want to go one step further (e.g. buffers of size 1000, 10000, 100000 and 10000000 (or whatever)). 
This could allow us to be slightly more efficient, but maybe I'm actually overengineering it and what you're proposing here is actually enough.

p0lyn0mial replied:
it is hard to say, I think we would have to benchmark it to see the value. I think we could start with the simple pool and refine it in the future.

wojtek-t replied:
I'm fine with starting simple - just maybe add a comment that something like that may be considered later.

p0lyn0mial commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go:
```diff
@@ -161,15 +162,25 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
 
+// EncodeWithAllocator writes an object to the provided writer.
+// In addition, it allows for providing a memory allocator for efficient memory usage during object serialization.
+func (s *Serializer) EncodeWithAllocator(obj runtime.Object, w io.Writer, memAlloc runtime.MemoryAllocator) error {
```
I will apply the same technique to the RawSerializer.

p0lyn0mial replied:
do you know why we can't switch the CRDs to proto?

wojtek-t replied:
For protobufs, we rely on auto-generated per-type code that is compiled in. For CRDs, we clearly can't reuse the same approach.
[We believe it's possible to support it, but it's a larger effort on its own.]

p0lyn0mial commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go:
```diff
@@ -161,15 +162,25 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
 }
 
-func (s *Serializer) doEncode(obj runtime.Object, w io.Writer) error {
+func (s *Serializer) doEncode(obj runtime.Object, w io.Writer, memAlloc runtime.MemoryAllocator) error {
```
it looks like we could optimize the `*runtime.Unknown:` case as well.

wojtek-t replied:
+1

p0lyn0mial committed:
fixes TestNestedEncodeError test

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/interfaces.go:
```diff
@@ -69,6 +69,22 @@ type Encoder interface {
 }
 
+// MemoryAllocator is responsible for allocating memory.
+// It exists to make the cost of object serialization cheaper. In some cases, it allows for allocating memory only once and then reusing it.
```
While the interface will be used for serialization, it's not tight to it.

Maybe just change to something like:

By encapsulating memory allocation into its own interface, we can reuse the memory
across many operations in places we know it can significantly improve the performance.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/interfaces.go:
```diff
@@ -69,6 +69,22 @@ type Encoder interface {
+}
+
+// EncoderAllocator  serializes objects in a way that allows callers to manage any additional memory allocations.
+type EncoderAllocator interface {
```
Call it EncoderWithAllocator maybe?

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/interfaces.go:
```diff
@@ -69,6 +69,22 @@ type Encoder interface {
+
+// EncoderAllocator  serializes objects in a way that allows callers to manage any additional memory allocations.
+type EncoderAllocator interface {
+	// EncodeWithAllocator writes an object to the provided writer.
```
How about explicitly nesting Encoder interface here too?

p0lyn0mial replied:
we could but it seems to be redundant atm

wojtek-t replied:
I think the Identifier() method is the important one for caching - and nesting it makes it more explicit.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/interfaces.go:
```diff
@@ -69,6 +69,22 @@ type Encoder interface {
+type EncoderAllocator interface {
+	// EncodeWithAllocator writes an object to the provided writer.
+	// In addition, it allows for providing a memory allocator for efficient memory usage during object serialization.
+	// This essentially allows for reusing memory instead of making additional allocations some serializers might require during serialization.
```
Maybe switch to sth like:

EncodeWithAllocator writes an object to a stream as Encode does.
However, it may utilize the passed MemoryAllocator to reduce the cost of memory allocations and
garbage collection.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/allocator.go:
```diff
@@ -0,0 +1,42 @@
+package serializer
```
boilerplate is missing (here and in other files)

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/allocator.go:
```diff
@@ -0,0 +1,42 @@
+// otherwise it returns previously allocated block of memory.
+//
+// Note that the returned array is not zeroed, it is the caller's
+// responsibility to clean the memory if needed.
```
That should go to an interface definition.

p0lyn0mial replied:
actually, it might depend on implementation, wdyt?

wojtek-t replied:
I would at least add at the interface level that implementation aren't obliged to zero the returned memory.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/allocator_test.go:
```diff
@@ -0,0 +1,71 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer/protobuf"
+)
+
```
Add some simple unit tests for the allocator (on top of the benchmarks below).

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/allocator_test.go:
```diff
@@ -0,0 +1,71 @@
+	"k8s.io/apimachinery/pkg/runtime/serializer/protobuf"
+)
+
+func BenchmarkProtobufEncoder(b *testing.B) {
```
Actually, these aren't really benchmarks of allocator, but rather benchmark of serializer.
Maybe move them to something like "encoder_with_allocator_test.go" or sth?

p0lyn0mial replied:
sure, I can move to a new file.

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:
```diff
@@ -206,6 +208,15 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	embeddedEncodeFn := s.EmbeddedEncoder.Encode
+	if encoder, supportsAllocator := s.EmbeddedEncoder.(runtime.EncoderAllocator); supportsAllocator {
+		memoryAllocator := serializer.AllocatorPool.Get().(*serializer.Allocator)
+		defer serializer.AllocatorPool.Put(memoryAllocator)
```
Should these two lines go into embeddedEncodeFn ?

p0lyn0mial replied:
I don't think so. We don't want to allocate/deallocate on every call to the encoding method. Instead, we allocate the buffer for the entire watch session and release it when we close the connection.

wojtek-t replied:
OK - that's fair point. Maybe just add a comment.

p0lyn0mial committed:
introduces a memory allocator

The primary use case for the allocator is to reduce cost of object serialization.
Initially it will be used by the protobuf serializer.
This approach puts less load on GC and leads to less fragmented memory in general.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go:
```diff
@@ -161,22 +162,37 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
 		estimatedSize := prefixSize + uint64(t.Size())
-		data := make([]byte, estimatedSize)
+		var data []byte
+		if memAlloc != nil {
```
Instead of this, can we add sth like `SimpleAllocator` which effectively will be always allocating the needed memory.

And just pass instead when non memAlloc is passed?

p0lyn0mial replied:
we could do that, would it be okay to create `SimpleAllocator` inside this `protobuf` package?

p0lyn0mial replied:
an alternative would be to create a new package `allocator` and move the `SimpleAllocator` and the `Allocator` there. Doing so would allow it to be reused.

wojtek-t replied:
I would personally put it in allocator.go that you're already adding.
It seems as simple as:

```
type SimpleAlocator struct{}{}

func (*SimpleAlocator) Allocate(n uint64) []byte {
  return make([]byte, n, n)
}
```

p0lyn0mial replied:
putting it into `allocator.go` would create an import cycle

p0lyn0mial replied:
creating a new package would prevent it

wojtek-t replied:
I don't understand - import cycle with what?

Allocator is not importing anything. If we don't have import cycle problem now, adding a second implementation of the same interface in the same file won't cause any new problems.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf_test.go:
```diff
@@ -66,3 +71,155 @@ func (t *mockTyper) ObjectKinds(obj runtime.Object) ([]schema.GroupVersionKind,
+		creater runtime.ObjectCreater
+		obj     runtime.Object
+	}{
+		// scenario 1
```
remove

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf_test.go:
```diff
@@ -66,3 +71,155 @@ func (t *mockTyper) ObjectKinds(obj runtime.Object) ([]schema.GroupVersionKind,
+			},
+		},
+
+		// scenario 2
```
remove (and empty line above)

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf_test.go:
```diff
@@ -66,3 +71,155 @@ func (t *mockTyper) ObjectKinds(obj runtime.Object) ([]schema.GroupVersionKind,
+			name:    "roundtrip a runtime.Unknown obj",
+			gvk:     schema.GroupVersionKind{Group: "group", Version: "version", Kind: "Unknown"},
+			creater: &mockCreater{obj: &runtime.Unknown{}},
+			obj:     &runtime.Unknown{TypeMeta: runtime.TypeMeta{APIVersion: "group/version", Kind: "Unknown"}},
```
Let's also set some Raw data in it.

p0lyn0mial replied:
it will make the test more complicated as we would have to compare the decoded object stored in the Raw data to something else that `obj`.  My intention was just to test various cases in the encoder. If we want to test the input perhaps we should add some fuzzer tests.

wojtek-t replied:
This is related to my other comment - I don't care about decoding.

We should be comparing whether the serialized version of data is the same (which mitigates your concern above).

p0lyn0mial replied:
it will work only if we remove the first Decoding call as well,  otherwise, it will fail.

wojtek-t replied:
Yes - I guess I wasn't explicit enough, but I would actually remove decoding completely.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf_test.go:
```diff
@@ -66,3 +71,155 @@ func (t *mockTyper) ObjectKinds(obj runtime.Object) ([]schema.GroupVersionKind,
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// test data
```
nit: remove

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf_test.go:
```diff
@@ -66,3 +71,155 @@ func (t *mockTyper) ObjectKinds(obj runtime.Object) ([]schema.GroupVersionKind,
+			if alloc.allocateCount != 1 {
+				t.Fatalf("expected the Allocate method to be called exactly 1 but it was executed: %v times ", alloc.allocateCount)
+			}
+			decodedObject, _, err = target.Decode(writer.Bytes(), &tc.gvk, nil)
```
Instead of testing Decoding again here, I would actually ensure that the serialized version is exactly the same.

So let's maybe test if the writer1.Bytes() are equal to writer2.Bytes()

p0lyn0mial replied:
are you afraid that the new encoding method would add some data and the `Decode` method would actually remove it?

wojtek-t replied:
So conceptually what we need is to ensure bckward-compatibility of the wire format (which is effectively a serialized data).

Decoder may ignore some data (and in fact IIRC it already drops unknown extensions). So if the encoder would be adding some well-structured data, it's completely possible that decoder would actually drop it.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go:
```diff
@@ -182,19 +182,40 @@ func (c *codec) Decode(data []byte, defaultGVK *schema.GroupVersionKind, into ru
+				return encoder.EncodeWithAllocator(obj, w, memAlloc)
+			}
+		} else {
+			// TODO: return an error or log and let it use the standard encode method
```
We don't want to fail - it's a legitimate case when the embedded encoder doesn't support it.

I wouldn't even log a warning - instead just maybe log some V(4) log or sth.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go:
```diff
@@ -182,19 +182,40 @@ func (c *codec) Decode(data []byte, defaultGVK *schema.GroupVersionKind, into ru
 	return out, gvk, strictDecodingErr
 }
 
+// EncodeWithAllocator ensures the provided object is output in the appropriate group and version, invoking
```
Add

var _ runtime.EncoderWithAllocator = &coder{}

somewhere above

wojtek-t commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go:
```diff
@@ -206,6 +208,17 @@ func (s *WatchServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 
+	embeddedEncodeFn := s.EmbeddedEncoder.Encode
+	if encoder, supportsAllocator := s.EmbeddedEncoder.(runtime.EncoderWithAllocator); supportsAllocator {
+		// note we don't put the memoryAllocator inside the embeddedEncodeFn method as that would allocate/deallocate memory on every call.
```
nit: can you shorten the lines?

wojtek-t reviewed (commented):
This is great - I added some comments but those are relatively small.

The only bigger one is missing support for JSON (for CRD purpose).

aojea commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/allocator.go:
```diff
@@ -0,0 +1,63 @@
+		return a.buf
+	}
+	// grow the buffer
+	// TODO: grow 2x the current size
```
should we use something like the logic used by the stdlib for buffers?  it does 2*cap(a.buf) + n

https://github.com/golang/go/blob/2ebe77a2fda1ee9ff6fd9a3e08933ad1ebaea039/src/bytes/buffer.go#L114-L150

if

p0lyn0mial replied:
I am fine with `2*cap(a.buf) + n` as well.

p0lyn0mial replied:
Also, I don't think at this point we need shrinking. We are bounded by the max size an object can have when persisted in etcd. In the worst case, the buffer should not exceed ~2MB

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/allocator.go:
```diff
@@ -0,0 +1,75 @@
+
+// SimpleAllocator a wrapper around make([]byte)
+// conforms to the MemoryAllocator interface
+// makes code utilizing it more elegant
```
nit: remove this line

p0lyn0mial commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/encoder_with_allocator_test.go:
```diff
@@ -0,0 +1,125 @@
+	for _, tc := range benchTestCases() {
+		b.Run(tc.name, func(b *testing.B) {
+			b.ReportAllocs()
+			target := protobuf.NewSerializer(nil, nil)
```
I want to add tests for the `RawSerializer` too

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf_test.go:
```diff
@@ -66,3 +70,115 @@ func (t *mockTyper) ObjectKinds(obj runtime.Object) ([]schema.GroupVersionKind,
+		t.Run(tc.name, func(t *testing.T) {
+			target := NewSerializer(nil, nil)
+
+			// step 1: perform enc/dec of the obj using target.Encode method
```
nit: remove this - the code is self-descriptive (same in other similar places)

p0lyn0mial commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf_test.go:
```diff
@@ -66,3 +70,115 @@ func (t *mockTyper) ObjectKinds(obj runtime.Object) ([]schema.GroupVersionKind,
+				t.Fatal(err)
+			}
+
+			// step 2: perform enc/dec of the obj using target.EncodeWithAllocator method
```
note for me to update the doc

wojtek-t reviewed (commented):
This is great - just some nits.

Also @p0lyn0mial - is this still WIP?

p0lyn0mial commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go:
```diff
@@ -182,19 +184,40 @@ func (c *codec) Decode(data []byte, defaultGVK *schema.GroupVersionKind, into ru
 }
 
-func (c *codec) doEncode(obj runtime.Object, w io.Writer) error {
+func (c *codec) doEncode(obj runtime.Object, w io.Writer, memAlloc runtime.MemoryAllocator) error {
```
I want to check`EncodeNestedObjects`, I need to make sure it doesn't require the memory allocator.

wojtek-t replied:
I can't see why would require, but adding an explicit test for it seems like a good AI

p0lyn0mial replied:
okay, I see, `NestedObjectEncoder` is an optional interface that objects may implement. As of today implemented only by `KubeSchedulerConfiguration` type.

p0lyn0mial commented:
> Also @p0lyn0mial - is this still WIP?

Two things from my side. Some benchmarks for the `RawSerializer` and making sure `EncodeNestedObjects` doesn't require memory allocator support. WDYT?

weilaaa commented on staging/src/k8s.io/apimachinery/pkg/runtime/allocator.go:
```diff
@@ -0,0 +1,75 @@
+// by caching created but unused items for later reuse, relieving pressure on the garbage collector.
+//
+// Usage:
+//  memoryAllocator := serializer.AllocatorPool.Get().(*serializer.Allocator)
```
Might be runtime.AllocatorPool.Get().(*serializer.Allocator)?

p0lyn0mial replied:
it it actually 
```
//  memoryAllocator := runtime.AllocatorPool.Get().(*runtime.Allocator)
//  defer runtime.AllocatorPool.Put(memoryAllocator)
```
:)

p0lyn0mial committed:
introduces a new streaming encoder that utilizes a memory allocator during objects serialization

p0lyn0mial commented:
ok, so I have addressed the recent comments, added some benchmarks and `NewEncoderWithAllocator` for the `RawSerializer`, PTAL.

p0lyn0mial commented:
/kind feature

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/encoder_with_allocator_test.go:
```diff
@@ -0,0 +1,87 @@
+			b.Fatal(err)
+		}
+	}
+}
```
Ping

p0lyn0mial replied:
heh, so my first idea was to create an HTML table to present the results. That turned out to be labor-intensive for me. So I wanted to use https://pkg.go.dev/golang.org/x/perf/cmd/benchstat but haven't found a good way to presents the results either. 

Would be okay to just copy/paste the results from my terminal?

wojtek-t replied:
Yes - I'm fine with anything as long as I can read it :)

p0lyn0mial replied:
done

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go:
```diff
@@ -161,22 +162,35 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
-func (s *Serializer) doEncode(obj runtime.Object, w io.Writer) error {
+func (s *Serializer) doEncode(obj runtime.Object, w io.Writer, memAlloc runtime.MemoryAllocator) error {
+	if memAlloc == nil {
+		return fmt.Errorf("a memory allocator must be provided")
```
I would prefer to logging an Error and falling back to SimpleAllocator in this case rather than failing the whole encoding.

p0lyn0mial replied:
it cannot happen. I added it as a safety net, `doEncode` is an internal method.

p0lyn0mial replied:
`memAlloc` is provided by `Encode` and `EncodeWithAllocator` methods.

wojtek-t replied:
Yes - an one can call EncodeWithAllocator and pass nil memoryallocator

p0lyn0mial replied:
In that case, a caller should use the `Encode` method or provide a noop allocator.
I would prefer to make it explicit as it is less error-prone, i.e. some middle layer doesn't pass the provided allocated down.

wojtek-t replied:
They should use `Encode` but they made a bug. I think it's still better to mitigate the problem and not fail the request on the cost of worse performance.

wojtek-t commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/protobuf/protobuf.go:
```diff
@@ -395,19 +409,32 @@ func unmarshalToObject(typer runtime.ObjectTyper, creater runtime.ObjectCreater,
-func (s *RawSerializer) doEncode(obj runtime.Object, w io.Writer) error {
+func (s *RawSerializer) doEncode(obj runtime.Object, w io.Writer, memAlloc runtime.MemoryAllocator) error {
+	if memAlloc == nil {
+		return fmt.Errorf("a memory allocator must be provided")
```
Same here

wojtek-t reviewed (commented):
Just some minor nits - other than that LGTM

wojtek-t commented:
I was actually expecting this, but to close the loop - the benchmark results are amazing!

wojtek-t commented:
/lgtm
/approve

Thanks!

wojtek-t commented:
/retest

wojtek-t commented:
/retest
